### ISSUE TITLE
unit test compile error Msvc  Release 

### ISSUE THREAD

**Issue by sunflower-knight** [alibaba/yalantinglibs#99]
### Search before asking

- [ ] I searched the [issues](https://github.com/alibaba/yalantinglibs/issues) and found no similar issues.
 
`MSVC+Release+ninja/MSBuild/NMake` have the same error
error C4737: 无法执行所需尾调用。性能可能会降级。

> WINDOW11 + VS 2022  
> HOST : WINDOW X64
> TARGET: WINDOW X64
> MSVC: CXX compiler identification is MSVC 19.34.31935.0
> SDK: Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22621.

### What happened + What you expected to happen

![image](https://user-images.githubusercontent.com/70421213/206052022-33d60b8c-a602-4b01-8344-5efe72937c67.png)

**Comment by poor-circle** [comment]
Strange error.

**Comment by qicosmos** [comment]
Could you use an elder version to test, before friendly cmake PR.

**Comment by sunflower-knight** [comment]
after some test I find that if you use `add_compile_options("/EHa")`  for `msvc` `release` `host-win64+target-win64`  , unit test code can compile and run . 
But maybe "/EHa" is not so good.  Here is a page talk about /E option for msvc https://learn.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-170 

---------------
I also find that  if you use win32 toolchain :  `msvc` `release` `host-x86+target-x86` unit test can compile, but they can't run.
`C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.34.31933/bin/Hostx86/x86/cl.exe`

![image](https://user-images.githubusercontent.com/70421213/206199766-4453dbbb-377e-449b-b9ff-f02bd4a0cdc6.png)

Strange..............................

**Comment by poor-circle** [comment]
> after some test I find that if you use `add_compile_options("/EHa")` for `msvc` `release` `host-win64+target-win64` , unit test code can compile and run . But maybe "/EHa" is not so good. Here is a page talk about /E option for msvc https://learn.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-170
> 
> I also find that if you use win32 toolchain : `msvc` `release` `host-x86+target-x86` unit test can compile, but they can't run. `C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.34.31933/bin/Hostx86/x86/cl.exe`
> 
> ![image](https://user-images.githubusercontent.com/70421213/206199766-4453dbbb-377e-449b-b9ff-f02bd4a0cdc6.png)
> 
> Strange..............................

Yes, I found this workaround too. And I found some exists  unfixed issue in msvc.It seems a bug in msvc tail recursion optimization.

**Comment by poor-circle** [comment]
See [bug report](https://developercommunity.visualstudio.com/t/coroutine-compilation-resulting-in-error-c4737-una/1510427)

**Comment by poor-circle** [comment]
Fix by #107
